### ISSUE TITLE
Malformed constraint related to monad transformers

### ISSUE THREAD

**Issue by rosekunkel** [ucsd-progsys/liquidhaskell#1568]
The following file produces a malformed constraint error:
```haskell
module MalformedTypeclass where

{-@ data IdentityT m a = IdentityT (m a) @-}
data IdentityT m a = IdentityT (m a)

class MyMonad m where
  myReturn :: a -> m a

{-@
instance MyMonad (IdentityT m) where
  myReturn :: a -> IdentityT m a
@-}
instance MyMonad (IdentityT m) where
  myReturn = undefined

-- When using the real Monad class the error doesn't happen until return is
-- actually used for some reason
{-@ testMyReturn :: a -> IdentityT m a @-}
testMyReturn :: a -> IdentityT m a
testMyReturn x = myReturn x
```

The error is
```
MalformedTypeclass.hs:14:3-10: Error: Uh oh.
  
 14 |   myReturn = undefined
        ^^^^^^^^
  
     addC: malformed constraint:
lq_tmp$x##780:{lq_tmp$x##1087 : a##a18a^"lq_tmp$x##1086" "lq_tmp$x##783" | $k_##1085[VV##1084:=lq_tmp$x##1087]
                                                                           && $k_##782[VV##781:=lq_tmp$x##1087]} -> {VV##790 : (MalformedTypeclass.IdentityT {lq_tmp$x##1083 : m##a186^"lq_tmp$x##1082" "lq_tmp$x##786" | $k_##1081[VV##1080:=lq_tmp$x##1083]
                                                                                                                                                                                                                          && $k_##785[VV##784:=lq_tmp$x##1083]} {lq_tmp$x##1087 : a##a18a^"lq_tmp$x##1086" "lq_tmp$x##789" | $k_##1085[VV##1084:=lq_tmp$x##1087]
                                                                                                                                                                                                                                                                                                                             && $k_##788[VV##787:=lq_tmp$x##1087]})^"lq_tmp$x##792" | $k_##791}
 <: 
lq_tmp$x##779:m##a186 -> (MalformedTypeclass.IdentityT a##a18a m##a186)
```

Stripping away the refinements, the malformed constraint is:
```
a -> IdentityT m a
<:
m -> IdentityT a m
```
so somehow the type arguments are getting switched.

**Comment by rosekunkel** [comment]
This seems to be the same error: #1311

**Comment by ranjitjhala** [comment]
Does the core contain cast or coerce?

**Comment by rosekunkel** [comment]
It has cast in it in some places:
```
[$cmyReturn
   :: forall (m :: * -> *) a. a -> MalformedTypeclass.IdentityT m a
 [LclId]
 $cmyReturn
   = \ (@ (m :: * -> *)) (@ a) ->
       let {
         $dIP_a2at :: GHC.Stack.Types.HasCallStack
         [LclId]
         $dIP_a2at
           = (GHC.Stack.Types.pushCallStack
                (GHC.CString.unpackCString# "undefined"#,
                 GHC.Stack.Types.SrcLoc
                   (GHC.CString.unpackCString# "main"#)
                   (GHC.CString.unpackCString# "MalformedTypeclass"#)
                   (GHC.CString.unpackCString#
                      "/home/rose/research/scratch/MalformedTypeclass.hs"#)
                   (GHC.Types.I# 14#)
                   (GHC.Types.I# 14#)
                   (GHC.Types.I# 14#)
                   (GHC.Types.I# 23#))
                ((GHC.Stack.Types.emptyCallStack
                  `cast` (Sym (GHC.Classes.N:IP[0]
                                   <"callStack">_N <GHC.Stack.Types.CallStack>_N)
                          :: GHC.Types.Coercible
                               GHC.Stack.Types.CallStack (?callStack::GHC.Stack.Types.CallStack)))
                 `cast` (GHC.Classes.N:IP[0]
                             <"callStack">_N <GHC.Stack.Types.CallStack>_N
                         :: GHC.Types.Coercible
                              (?callStack::GHC.Stack.Types.CallStack)
                              GHC.Stack.Types.CallStack)))
             `cast` (Sym (GHC.Classes.N:IP[0]
                              <"callStack">_N <GHC.Stack.Types.CallStack>_N)
                     :: GHC.Types.Coercible
                          GHC.Stack.Types.CallStack
                          (?callStack::GHC.Stack.Types.CallStack)) } in
       break<1>()
       src</home/rose/research/scratch/MalformedTypeclass.hs:14:3-22>
       GHC.Err.undefined
         @ 'GHC.Types.LiftedRep
         @ (a -> MalformedTypeclass.IdentityT m a)
         $dIP_a2at,
 MalformedTypeclass.$fMyMonadIdentityT [InlPrag=INLINE (sat-args=0)]
   :: forall (m :: * -> *).
      MalformedTypeclass.MyMonad (MalformedTypeclass.IdentityT m)
 [LclIdX[DFunId(nt)],
  Unf=Unf{Src=InlineStable, TopLvl=True, Value=False, ConLike=False,
          WorkFree=True, Expandable=True,
          Guidance=ALWAYS_IF(arity=0,unsat_ok=False,boring_ok=True)
          Tmpl= \ (@ (m :: * -> *)) ->
                  ($cmyReturn @ m)
                  `cast` (Sym (MalformedTypeclass.N:MyMonad[0]
                                   <MalformedTypeclass.IdentityT m>_N)
                          :: GHC.Types.Coercible
                               (forall a. a -> MalformedTypeclass.IdentityT m a)
                               (MalformedTypeclass.MyMonad (MalformedTypeclass.IdentityT m)))}]
 MalformedTypeclass.$fMyMonadIdentityT
   = \ (@ (m :: * -> *)) ->
       ($cmyReturn @ m)
       `cast` (Sym (MalformedTypeclass.N:MyMonad[0]
                        <MalformedTypeclass.IdentityT m>_N)
               :: GHC.Types.Coercible
                    (forall a. a -> MalformedTypeclass.IdentityT m a)
                    (MalformedTypeclass.MyMonad (MalformedTypeclass.IdentityT m))),
 MalformedTypeclass.testMyReturn
   :: forall a (m :: * -> *). a -> MalformedTypeclass.IdentityT m a
 [LclIdX]
 MalformedTypeclass.testMyReturn
   = \ (@ a) (@ (m :: * -> *)) (x :: a) ->
       break<0>(x)
       src</home/rose/research/scratch/MalformedTypeclass.hs:20:1-27>
       MalformedTypeclass.myReturn
         @ (MalformedTypeclass.IdentityT m)
         (MalformedTypeclass.$fMyMonadIdentityT @ m)
         @ a
         (src</home/rose/research/scratch/MalformedTypeclass.hs:20:27> x),
 MalformedTypeclass.$trModule :: GHC.Types.Module
 [LclIdX]
 MalformedTypeclass.$trModule
   = GHC.Types.Module
       (GHC.Types.TrNameS "main"#)
       (GHC.Types.TrNameS "MalformedTypeclass"#),
 $krep_a2bW [InlPrag=NOUSERINLINE[~]] :: GHC.Types.KindRep
 [LclId]
 $krep_a2bW
   = GHC.Types.KindRepTyConApp
       GHC.Types.$tcConstraint (GHC.Types.[] @ GHC.Types.KindRep),
 $krep_a2bV [InlPrag=NOUSERINLINE[~]] :: GHC.Types.KindRep
 [LclId]
 $krep_a2bV = GHC.Types.KindRepFun GHC.Types.krep$*Arr* $krep_a2bW,
 $krep_a2bX [InlPrag=NOUSERINLINE[~]] :: GHC.Types.KindRep
 [LclId]
 $krep_a2bX
   = GHC.Types.KindRepFun GHC.Types.krep$*Arr* GHC.Types.krep$*Arr*,
 $krep_a2c1 [InlPrag=NOUSERINLINE[~]] :: GHC.Types.KindRep
 [LclId]
 $krep_a2c1 = GHC.Types.$WKindRepVar (GHC.Types.I# 1#),
 $krep_a2c0 [InlPrag=NOUSERINLINE[~]] :: GHC.Types.KindRep
 [LclId]
 $krep_a2c0 = GHC.Types.$WKindRepVar (GHC.Types.I# 0#),
 $krep_a2bZ [InlPrag=NOUSERINLINE[~]] :: GHC.Types.KindRep
 [LclId]
 $krep_a2bZ = GHC.Types.KindRepApp $krep_a2c0 $krep_a2c1,
 MalformedTypeclass.$tcIdentityT :: GHC.Types.TyCon
 [LclIdX]
 MalformedTypeclass.$tcIdentityT
   = GHC.Types.TyCon
       12410962577597022282##
       15736914772789928699##
       MalformedTypeclass.$trModule
       (GHC.Types.TrNameS "IdentityT"#)
       0#
       $krep_a2bX,
 $krep_a2c2 [InlPrag=NOUSERINLINE[~]] :: GHC.Types.KindRep
 [LclId]
 $krep_a2c2
   = GHC.Types.KindRepTyConApp
       MalformedTypeclass.$tcIdentityT
       (GHC.Types.:
          @ GHC.Types.KindRep
          $krep_a2c0
          (GHC.Types.:
             @ GHC.Types.KindRep
             $krep_a2c1
             (GHC.Types.[] @ GHC.Types.KindRep))),
 $krep_a2bY [InlPrag=NOUSERINLINE[~]] :: GHC.Types.KindRep
 [LclId]
 $krep_a2bY = GHC.Types.KindRepFun $krep_a2bZ $krep_a2c2,
 MalformedTypeclass.$tc'IdentityT :: GHC.Types.TyCon
 [LclIdX]
 MalformedTypeclass.$tc'IdentityT
   = GHC.Types.TyCon
       4974790856926696172##
       8865973219025067224##
       MalformedTypeclass.$trModule
       (GHC.Types.TrNameS "'IdentityT"#)
       2#
       $krep_a2bY,
 MalformedTypeclass.$tcMyMonad :: GHC.Types.TyCon
 [LclIdX]
 MalformedTypeclass.$tcMyMonad
   = GHC.Types.TyCon
       16051466632236452693##
       8150582087488838184##
       MalformedTypeclass.$trModule
       (GHC.Types.TrNameS "MyMonad"#)
       0#
       $krep_a2bV]
```

**Comment by nikivazou** [comment]
It is a bug, I am into it. 
The above "fix" moves the error from the instance definition to its call side.

**Comment by ranjitjhala** [comment]
Thanks @nikivazou!

**Comment by nikivazou** [comment]
@rosekunkel  can you check all is good?

**Comment by rosekunkel** [comment]
This problem seems resolved, but now I've uncovered a new one: #1571
